### PR TITLE
Cache kmp

### PR DESCRIPTION
**Pull request by aclassen**
Please see our contributing guidelines (contributing.md) primarily make sure to sign our cla as we cannot accept code externally without a signed cla

https://opensource.dropbox.com/cla/

### CODE REVIEW

File: app/src/main/AndroidManifest.xml
Comment (digitalbuddha) [inline app/src/main/AndroidManifest.xml:14]: thanks!

File: cache/build.gradle
Comment (chris-mitchell) [inline cache/build.gradle:25]: Can we keep Jacoco and kmp or is it not supported?

File: buildsystem/dependencies.gradle
Comment (digitalbuddha) [inline buildsystem/dependencies.gradle:6]: checking internally what this does to our tooling...will post answer shortly

File: buildsystem/dependencies.gradle
Comment (digitalbuddha) [inline buildsystem/dependencies.gradle:29]: same need to validate/upgrade things internally but this won't block the diff

File: cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/Cache.kt
Comment (chris-mitchell) [inline cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/Cache.kt:2]: Is it an issue these are under a copyright of the author or should we move it to Dropbox?

File: cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/LocalCache.kt
Comment (chris-mitchell) [inline cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/LocalCache.kt:1032]: Should this come before the `unlock` call above? I (think) in the call hierarchy we ultimately call `runLockedCleanup` which gets the lock again which feels like we could get into some race conditions.  Thoughts?

File: cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/LocalCache.kt
Comment (chris-mitchell) [inline cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/LocalCache.kt:868]: nit: unnecessary comment?

File: cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/LocalCache.kt
Comment (chris-mitchell) [inline cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/LocalCache.kt:1374]: nit: unnecessary method?

File: cache/src/commonTest/kotlin/com/dropbox/kmp/external/cache3/DefaultCacheTest.kt
Comment (chris-mitchell) [inline cache/src/commonTest/kotlin/com/dropbox/kmp/external/cache3/DefaultCacheTest.kt:8]: Thanks for all the tests!

File: cache/src/commonTest/kotlin/com/dropbox/kmp/external/cache3/TestData.kt
Comment (digitalbuddha) [inline cache/src/commonTest/kotlin/com/dropbox/kmp/external/cache3/TestData.kt:3]: nit: please change to non data class as data classes are not binary compatible when making changes https://jakewharton.com/public-api-challenges-in-kotlin/

File: cache/build.gradle
Comment (digitalbuddha) [reply cache/build.gradle:25]: I'll take care of this one after the fact, implementor had trouble with jacoco and version changes. This is not going into main so no concern for now

File: cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/Cache.kt
Comment (digitalbuddha) [reply cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/Cache.kt:2]: Nope! P-Counsel signed off on this. As long as contributor signs our CLA which states the code is theirs and not stolen + we can use with our license. The copyright is similar with what we do with Googlers (see Multicaster)

**Review by rharter — CHANGES_REQUESTED** [review]
So I noticed as I was trying to verify this works locally that the other modules in this repo aren't actually using this dependency due to [this](https://github.com/dropbox/Store/blob/main/buildsystem/dependencies.gradle#L57-L58),  so we can't run tests to verify that this updated implementation provides what Store needs.

I realize that is out of scope of "Updating Cache to KMP", but I would say is a blocker to be able to merge a change this significant.  Could you please update the other modules to use the cache module, then update their call-sites accordingly to ensure that the new implementations is compatible?

File: cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/Cache.kt
Comment (rharter) [reply cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/Cache.kt:2]: There is an open question to Legal of this being a port of the Guava cache implementation and not retaining their original copyright. I noticed the LocalCache implementation does that, but this is also derived from there (perhaps indirectly, but it's apparent from the JavaDocs) so I would it would need to retain the "Guava Authors" attribution.

Comment (aclassen) [reply cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/Cache.kt:2]: I used the sources from :
https://github.com/dropbox/Store/tree/main/cache/src/main/java/com/dropbox/android/external/cache3
Where no copyright is in the source files.
So i added the copyright from the guava sources , as it's not my intention at all to claim the guava authors work.
I'll change this in the other files too, i just missed it.

**Comment by ben-manes** [comment]
fwiw, you might use [Lincheck](https://www.google.com/search?q=lincheck&oq=lincheck&aqs=chrome..69i57j0i10l4j69i60j69i61l2.2713j0j4&sourceid=chrome&ie=UTF-8) ([example](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/test/java/com/github/benmanes/caffeine/cache/testing/AbstractLincheckMapTest.java)) for some concurrency testing. Since we didn't design Guava's cache for Android (where [LruCache](https://developer.android.com/reference/android/util/LruCache) was preferred), you might prefer a simpler approach, e.g. LruCache + futures, over maintaining a port of that code base.

File: cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/Cache.kt
Comment (rharter) [reply cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/Cache.kt]: Thanks, I figured that was the case (and pointed out to our legal counsel that our previous version didn't retain the license in the source).

**Comment by rharter** [comment]
> So I noticed as I was trying to verify this works locally that the other modules in this repo aren't actually using this dependency due to [this](https://github.com/dropbox/Store/blob/main/buildsystem/dependencies.gradle#L57-L58), so we can't run tests to verify that this updated implementation provides what Store needs.
> 
> I realize that is out of scope of "Updating Cache to KMP", but I would say is a blocker to be able to merge a change this significant. Could you please update the other modules to use the cache module, then update their call-sites accordingly to ensure that the new implementations is compatible?

I should point out that I understand this isn't the PR author's mistake or anything, just an unfortunate artifact of history that prevents the proper testing (or use) of this change.

**Comment by lukisk** [comment]
Hi Andre, why was this PR closed ?

**Comment by aclassen** [comment]
As there was no response/interest anymore, i decided to close it and continue in my private repo.

**Comment by lukisk** [comment]
:(, do you have any plans to publish it? We use it(manually merged your cache branch into kmp..) in android/ios apps and everything works fine. If you want, i can participate ..

**Comment by aclassen** [comment]
You can find it in my fork

File: buildsystem/dependencies.gradle
Comment (digitalbuddha) [reply buildsystem/dependencies.gradle:6]: All set here as no more dropbox ownership

File: cache/api/cache.api
Comment (digitalbuddha) [inline cache/api/cache.api:3]: Curious what is the new api used for? I'm always hesitant to support more public functionality

File: cache/build.gradle
Comment (digitalbuddha) [reply cache/build.gradle:29]: Let's remove jacaoco for now and make a ticket to add back in  please

**Review by digitalbuddha — COMMENTED** [review]
I think this is very close. I'd like to review once more this weekend and then we can merge. Thank you for the incredible effort

File: cache/api/cache.api
Comment (aclassen) [reply cache/api/cache.api:3]: I introduced this because:
`public abstract fun get (Ljava/lang/Object;Ljava/util/concurrent/Callable;)Ljava/lang/Object;`
is not implemented yet.

**Review by digitalbuddha — APPROVED** [review]
I'm good to see this merge if you are. Anything else you wanted to add? 

I imagine some can continue to use an older cache artifact if they want. When releasing maybe we can rename the version to something with -kmp to indicate new cache

**Comment by aclassen** [comment]
Yes, i think renaming would be an good idea.
I would only update to kotlin 1.7.0 and corresponding dependencies, but we can do this also later.

**Comment by aclassen** [comment]
Ready to be merged 👍

**Comment by digitalbuddha** [comment]
One step closer to store kmp thank you for all your hard work

**Comment by aclassen** [comment]
I'll start to prepare the next one

**Pull request merged**
